[libplctag] Add new port

gotnone opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

Commit:
[libplctag] Add new port (#33679)

gotnone commented:
Upstream does not seem to support the *-uwp targets.

gotnone commented:
I think the failure on arm64_android may have been a false positive because the arm_neon_android was successful.  The error logs indicate that `version.h.in` did not exist resulting in a the failure 
````
CMake Error at CMakeLists.txt:335 (CONFIGURE_FILE):
  No such file or directory
````

Can anyone either rerun the pipeline or dig into the buildtree for the arm64_android build to see what may be causing this file to be deleted / not properly extracted from the source tarball?

dg0yt commented on ports/libplctag/portfile.cmake:
```diff
@@ -0,0 +1,72 @@
+	sanitizer.patch
+)
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
```
This may need 
```suggestion
vcpkg_cmake_configure(
    SOURCE_PATH "${SOURCE_PATH}"
    DISABLE_PARALLEL_CONFIGURE
```
because 
https://github.com/libplctag/libplctag/blob/4793d92893906e80fcfc67d4c4ed87fd37d81df8/CMakeLists.txt#L327 
https://github.com/libplctag/libplctag/blob/4793d92893906e80fcfc67d4c4ed87fd37d81df8/CMakeLists.txt#L544 
may write to the source directory during configuration.

gotnone replied:
Good catch.  That may explain the spurious build failure I saw earlier on arm64_android.

JonLiu1993 commented on ports/libplctag/portfile.cmake:
```diff
@@ -0,0 +1,72 @@
+vcpkg_cmake_install()
+vcpkg_fixup_pkgconfig()
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+	if(WIN32)
+		file(RENAME
+			"${CURRENT_PACKAGES_DIR}/lib/${VCPKG_TARGET_STATIC_LIBRARY_PREFIX}plctag_static${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}"
+			"${CURRENT_PACKAGES_DIR}/lib/${VCPKG_TARGET_STATIC_LIBRARY_PREFIX}plctag${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}")
+		file(RENAME
+			"${CURRENT_PACKAGES_DIR}/debug/lib/${VCPKG_TARGET_STATIC_LIBRARY_PREFIX}plctag_static${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}"
+			"${CURRENT_PACKAGES_DIR}/debug/lib/${VCPKG_TARGET_STATIC_LIBRARY_PREFIX}plctag${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}")
+	endif()
+	file(REMOVE
+		"${CURRENT_PACKAGES_DIR}/lib/${VCPKG_TARGET_SHARED_LIBRARY_PREFIX}plctag${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}"
+		"${CURRENT_PACKAGES_DIR}/lib/${VCPKG_TARGET_IMPORT_LIBRARY_PREFIX}plctag${VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX}"
+		"${CURRENT_PACKAGES_DIR}/debug/lib/${VCPKG_TARGET_SHARED_LIBRARY_PREFIX}plctag${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}"
+		"${CURRENT_PACKAGES_DIR}/debug/lib/${VCPKG_TARGET_IMPORT_LIBRARY_PREFIX}plctag${VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX}")
+else()
+	if(WIN32)
+		file(MAKE_DIRECTORY
+			"${CURRENT_PACKAGES_DIR}/bin"
+			"${CURRENT_PACKAGES_DIR}/debug/bin")
+		file(RENAME
+			"${CURRENT_PACKAGES_DIR}/lib/${VCPKG_TARGET_SHARED_LIBRARY_PREFIX}plctag${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}"
+			"${CURRENT_PACKAGES_DIR}/bin/${VCPKG_TARGET_SHARED_LIBRARY_PREFIX}plctag${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}")
+		file(RENAME
+			"${CURRENT_PACKAGES_DIR}/debug/lib/${VCPKG_TARGET_SHARED_LIBRARY_PREFIX}plctag${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}"
+			"${CURRENT_PACKAGES_DIR}/debug/bin/${VCPKG_TARGET_SHARED_LIBRARY_PREFIX}plctag${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}")
+		file(REMOVE
+			"${CURRENT_PACKAGES_DIR}/lib/${VCPKG_TARGET_STATIC_LIBRARY_PREFIX}plctag_static${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}"
+			"${CURRENT_PACKAGES_DIR}/debug/lib/${VCPKG_TARGET_STATIC_LIBRARY_PREFIX}plctag_static${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}")
+	else()
+		file(REMOVE
+			"${CURRENT_PACKAGES_DIR}/lib/${VCPKG_TARGET_STATIC_LIBRARY_PREFIX}plctag${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}"
+			"${CURRENT_PACKAGES_DIR}/debug/lib/${VCPKG_TARGET_STATIC_LIBRARY_PREFIX}plctag${VCPKG_TARGET_STATIC_LIBRARY_SUFFIX}")
+	endif()
+endif()
```
@gotnone, Thanks for your PR, The code here seems to be very redundant. What have been deleted here?

gotnone replied:
Problem 1: Upstream `CMakeLists.txt` automatically builds both static and dynamic libraries.

Problem 2: Under triplets where `WIN32` is set, there are additional complications that need to be handled.  Because upstream builds both types of libraries they needed to rename the static libraries so that they don't conflict with the import library.  Upstream handled this name conflict by inserting `_static` between the library base name and the LIBRARY_SUFFIX}

Upstream CMakeLists.txt L414:
```
    410 elseif(WIN32)
    411     # static library
    412         add_library(plctag_static STATIC ${libplctag_SRCS} )
    413         set_target_properties(plctag_static PROPERTIES LINK_FLAGS "${BASE_LINK_FLAGS}")
    414         set_target_properties(plctag_static PROPERTIES VERSION "${libplctag_VERSION_MAJOR}.${libplctag_VERSION_MINOR}" OUTPUT_NAME "plctag_static")
    415
    416     target_compile_definitions(plctag_dyn PUBLIC -DLIBPLCTAGDLL_EXPORTS=1)
    417         set(tool_lib "plctag_dyn")
    418 endif()
```

Problem 3: Upstream currently installs `WIN32` dlls into the `lib` directory when they need to be installed in `bin`

Upstream CMakeLists.txt L615:
```
    614 elseif(WIN32)
    615         install(TARGETS plctag_dyn DESTINATION lib${LIB_SUFFIX})
    616         install(TARGETS plctag_static DESTINATION lib${LIB_SUFFIX})
    617 endif()
```

Currently I am cleaning up these files after the fact with the commands in question.

Is it a better practice to try to further patch the upstream CMakeLists.txt to build static and dynamic libraries separately?

JonLiu1993 replied:
Thanks for your explanation, if there are more changes in the upstream CMakeLists I think it would be better for us to keep the current changes.

dg0yt replied:
> Is it a better practice to try to further patch the upstream CMakeLists.txt to build static and dynamic libraries separately?

Yes. It saves build time.

In the case of installation locations, moving after install may also create problems with the exported CMake config.

gotnone replied:
@dg0yt Is there a recommended set of best practices to split an upstream CMake build which produces both static & shared targets?

I was thinking that the process would look like the following

1. In a patch of the upstream `CMakeLists.txt`, wrap the two targets with `if(LIBPLCTAG_STATIC_BUILD)` and `if(LIBPLCTAG_SHARED_BUILD)`, respectively
  - Fix the upstream static library naming scheme

2. In `portfile.cmake` conditionally set these CMake variables for the appropriate triplet type
```
if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
    set(LIBPLCTAG_STATIC_BUILD=1)
else()
    set(LIBPLCTAG_SHARED_BUILD=1)
endif()
```

3. Fix any differences in the PkgConfig .pc file that are unique for static vs shared targets

gotnone replied:
It seems that the LIBPLCTAG_*_BUILD variables need to be passed from the portfile to the CMake configure process using the OPTIONS section of vcpkg_cmake_configure().

JonLiu1993 commented on ports/libplctag/portfile.cmake:
```diff
@@ -0,0 +1,72 @@
+endif()
+
+# Handle copyright
+file(READ "${SOURCE_PATH}/LICENSE-1.MPL" mpl)
+file(READ "${SOURCE_PATH}/LICENSE-2.LGPL" lgpl)
+file(WRITE "${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright"
+"${PORT} is released under a dual MPL / LGPL license.
```
```suggestion
vcpkg_install_copyright("${SOURCE_PATH}/LICENSE-1.MPL" ${SOURCE_PATH}/LICENSE-2.LGPL" COMMENT "LIBPLCTAG is released under a dual MPL / LGPL license")
```

gotnone committed:
Update ports/libplctag/portfile.cmake

Upstream CMakeLists.txt writes version.h to source directory with CONFIGURE_FILE command

Co-authored-by: Kai Pastor <dg0yt@darc.de>

gotnone committed:
Update ports/libplctag/portfile.cmake

Use vcpkg_install_copyright()

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

Commit:
Update CI baseline

Commit:
Clean up pkgconfig.patch

* Update CI baseline

Commit:
Fix manifest

* Limit pthreads dependency to !windows platforms

* Update CI baseline

JonLiu1993 commented:
Tested usage successfully by `libplctag:x64-windows`:
````
The package libplctag can be imported via CMake FindPkgConfig module:

    find_package(PkgConfig REQUIRED)
    pkg_check_modules(libplctag REQUIRED IMPORTED_TARGET libplctag)
    target_link_libraries(main PRIVATE PkgConfig::libplctag)
````

Commit:
Refactor cmake.patch

* Use BUILD_SHARED_LIBS to detect shared vs static builds

* Update CI baseline

BillyONeal commented on ports/libplctag/usage:
```diff
@@ -0,0 +1,5 @@
+The package libplctag can be imported via CMake FindPkgConfig module:
+
+    find_package(PkgConfig REQUIRED)
```
They seem to have `install(TARGETS` if your patches are any indication; is there a reason we are providing pkgconfig instructions rather than cmake targets instructions given that this is natively a CMake build system?

BillyONeal commented on ports/libplctag/portfile.cmake:
```diff
@@ -0,0 +1,29 @@
+  SHA512 f540780642112483bd199995b327342bd2df2ec4a4d44c2356886ab94879aeb2da3e420760da2f00b5aee0d1278fc582f73cba8227b276787d21b18fa12a3add
+  HEAD_REF release
+  PATCHES
+	cmake.patch
+	sanitizer.patch
+	pkgconfig.patch
+	fix_static.patch
+)
```
This is kind of a lot of patches for a brand new port; has upstream been informed of these changes?

BillyONeal replied:
(Some of the 'only build static or dynamic' is somewhat common but there are also like 'how dynamic is at all' changes in here)

BillyONeal commented on ports/libplctag/cmake.patch:
```diff
@@ -0,0 +1,120 @@
+ set_target_properties(plctag_dyn PROPERTIES SOVERSION "${libplctag_VERSION_MAJOR}.${libplctag_VERSION_MINOR}" OUTPUT_NAME "plctag")
++    set(tool_lib "plctag_dyn")
++    if(WIN32)
++        target_compile_definitions(plctag_dyn PUBLIC -DLIBPLCTAGDLL_EXPORTS=1)
```
This should almost certainly not be `PUBLIC`; that would be telling people who link with this to define `LIBPLCTAGDLL_EXPORTS`.

gotnone replied:
@dg0yt, I don't entirely understand the issue.  Is this something that we need to push upstream?

Will this actually affect users of this port?  As far as I understand they will be consuming the library using the pkgconfig file and not using FindPkg.

dg0yt replied:
It may not reach users now, but as soon this (CMake!) project starts to provide exported CMake config.

Given your example of
~~~
-    #ifdef LIBPLCTAGDLL_EXPORTS
+    #ifdef LIBPLCTAG_STATIC
+        #define LIB_EXPORT extern
+    #elif defined(LIBPLCTAGDLL_EXPORTS)
         #define LIB_EXPORT __declspec(dllexport)
     #else
         #define LIB_EXPORT __declspec(dllimport)
~~~
we see that `LIBPLCTAGDLL_EXPORTS` is for reaching `dllexport`.  Which is only needed during the build (PRIVATE), but never during usage. So it is simply wrong to use `PUBLIC`.

BillyONeal commented on ports/libplctag/cmake.patch:
```diff
@@ -0,0 +1,120 @@
++    endif()
+ endif()
+ 
+-if(ANDROID_BUILD)
++if(ANDROID_BUILD AND 0)
```
```suggestion
-if(ANDROID_BUILD)
+if(0)
```

BillyONeal reviewed (changes requested):
I'm marking this as draft; please feel to re-mark ready for review after looking at comments above. Thanks for the new port!

JonLiu1993 commented:
Ping @gotnone for response.

gotnone commented:
> Ping @gotnone for response.

Yes?  Has anyone had a chance to review my responses to @BillyONeal?

dg0yt commented:
> Yes? Has anyone had a chance to review my responses

@gotnone Did you submit them at all? Nothing visible here.

gotnone commented:
> > Yes? Has anyone had a chance to review my responses
> 
> @gotnone Did you submit them at all? Nothing visible here.

I have replied to the conversations regarding requested changes.  I received no response to my replies.  Is there something I need to do on my end to push or publish the reply?

![image](https://github.com/microsoft/vcpkg/assets/843070/a97c39be-d13b-48c8-a6a9-54ded118587b)

dg0yt commented:
Note the "Pending" - you are really still in the state of entering review comments. You still have to submit the comments using the "Review changes" button which is in the top-right region of the "Files changed" tab.

gotnone commented:
> Note the "Pending" - you are really still in the state of entering review comments. You still have to submit the comments using the "Review changes" button which is in the top-right region of the "Files changed" tab.

There doesn't seem to be anything up by the files changed tab on my browser that mentions review.
![image](https://github.com/microsoft/vcpkg/assets/843070/8feee4e2-2750-426a-9654-458aae592432)

The only button I see that mentions the word review is "Ready for review" located at the bottom of the PR conversation.  Is that what you are talking about?
![image](https://github.com/microsoft/vcpkg/assets/843070/4db1b036-54bf-4df3-87ae-568988726ab9)

dg0yt commented on ports/libplctag/pkgconfig.patch:
```diff
@@ -0,0 +1,13 @@
+ Version: @VERSION@
+ 
+-Libs: -L${libdir} -lplctag -pthread
++Libs: -L${libdir} -lplctag @PTHREAD_LDFLAGS@ @WINSOCK_LDFLAGS@
```
Is `@WINSOCK_LDFLAGS@` a candidate for `Libs.private`?

gotnone replied:
Is my understanding correct that Libs.private is used for static builds?

Is linking with -lws2_32 only needed during a static build?

dg0yt replied:
Yes, `Libs.private:` is expressing linking requirements for static usage.
I assume that linking ws2_32 is sufficiently covered by building libpcltag with dynamic linkage. When building with static linkage, it becomes a usage requirement because the linking takes place at the  downstream end of the chain.

(In addition, `vcpkg_fixup_pkconfig()` appends `Libs.private` to `Libs` when building with static linkage. That's why consumers don't need to use the `--static` option to `pkg-config`. It is simply resolved at package build time.)

But I didn't test/review this particular package.

dg0yt commented on ports/libplctag/pkgconfig.patch:
```diff
@@ -0,0 +1,13 @@
+-Libs: -L${libdir} -lplctag -pthread
++Libs: -L${libdir} -lplctag @PTHREAD_LDFLAGS@ @WINSOCK_LDFLAGS@
+ Cflags: -I${includedir}
++Cflags.private: -DLIBPLCTAG_STATIC
```
What does this macro control? For some purposes, it is more robust to replace its (few) occurrences in the headers.

gotnone replied:
Upstream static build was broken on Windows, in that the resulting library was not actually able to be linked.  Upstream used the troublesome `LIBPLCTAGDLL_EXPORTS` to switch between dllexport and dllimport decorators, but did not offer an undecorated or extern decorated option for consuming the API as a static library on windows.  There is an open issue upstream libplctag/libplctag#435.

See fix_static.patch:L18
```
-    #ifdef LIBPLCTAGDLL_EXPORTS
+    #ifdef LIBPLCTAG_STATIC
+        #define LIB_EXPORT extern
+    #elif defined(LIBPLCTAGDLL_EXPORTS)
         #define LIB_EXPORT __declspec(dllexport)
     #else
         #define LIB_EXPORT __declspec(dllimport)
```

dg0yt replied:
A familiar pattern...
This is the use case where replacing makes sense *for vcpkg*: Given that the result is known at build time, it could be hard-coded in installed headers. There are always users which don't use pkg-config...

Apart from that, it might make sense to prepare (some of your) patches at PR for upstream, and in particular to leverage CMakes standard capability to generate a header file for the DLL export stuff.

gotnone replied:
For something this simple, does it make sense to put the patch in `portfile.cmake`

```
if (VCPKG_LIBRARY_LINKAGE STREQUAL "static")
    file(READ "${SOURCE_PATH}/src/lib/libplctag.h" libplctag_h)
    file(WRITE "${CURRENT_PACKAGES_DIR}/include/libplctag.h"
"#define LIBPLCTAG_STATIC
${libplctag_h}
" )
endif()
```

If you were thinking something different, can you point to an example port that does conditional patching based on the type of build?

dg0yt replied:
I'm not talking about patching before the build. The port may need it anyways.

What you will find in a number of ports, e.g. zlib, is modifying the installed headers.
~~~cmake
if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/include/libplctag.h" "ifdef LIBPLCTAG_STATIC" "if 1")    
endif()
~~~

gotnone commented on ports/libplctag/sanitizer.patch:
```diff
@@ -0,0 +1,36 @@
+         SET(STR_OP_OVERFLOW_OFF "")
+     endif()
+ 
+-    set(BASE_RELEASE_FLAGS "${CMAKE_C_FLAGS} -Wall -pedantic -Wextra -Wconversion -fms-extensions -fno-strict-aliasing -fvisibility=hidden ${STR_OP_OVERFLOW_OFF} -D__USE_POSIX=1 -D_POSIX_C_SOURCE=200809L")
+-    set(BASE_DEBUG_FLAGS "${CMAKE_C_FLAGS} -g -Wall -pedantic -Wextra -Wconversion -fms-extensions -fno-strict-aliasing -fvisibility=hidden -fsanitize=address -fsanitize=leak -fsanitize=undefined ${STR_OP_OVERFLOW_ON} -D__USE_POSIX=1 -D_POSIX_C_SOURCE=200809L")
+-	set(C99_FLAGS "-std=c99" )
+-    set(BASE_RELEASE_LINK_FLAGS "")
+-    set(BASE_DEBUG_LINK_FLAGS "-fsanitize=address -fsanitize=leak -fsanitize=undefined")
++    SET(BASE_RELEASE_FLAGS "${CMAKE_C_FLAGS} -Wall -pedantic -Wextra -Wconversion -fms-extensions -fno-strict-aliasing -fvisibility=hidden ${STR_OP_OVERFLOW_OFF} -D__USE_POSIX=1 -D_POSIX_C_SOURCE=200809L")
++    SET(BASE_DEBUG_FLAGS "${CMAKE_C_FLAGS} -g -Wall -pedantic -Wextra -Wconversion -fms-extensions -fno-strict-aliasing -fvisibility=hidden ${STR_OP_OVERFLOW_ON} -D__USE_POSIX=1 -D_POSIX_C_SOURCE=200809L")
++	SET(C99_FLAGS "-std=c99" )
++    SET(BASE_RELEASE_LINK_FLAGS "")
++    SET(BASE_DEBUG_LINK_FLAGS "")
++
++    if(USE_SANITIZERS)
++        SET(SANITIZERS_FLAGS " -fsanitize=address -fsanitize=leak -fsanitize=undefined")
++        STRING(APPEND BASE_DEBUG_FLAGS ${SANITIZERS_FLAGS})
++        STRING(APPEND BASE_DEBUG_LINK_FLAGS ${SANITIZERS_FLAGS})
++    endif()
```
I pulled this directly from upstream's patchset.  All I know is that without it debug builds force the use of sanitizers and so I couldn't get this port to build without this patch.  Using this patch, sanitizers can be disabled with `-DUSE_SANITIZERS=0`

https://github.com/libplctag/libplctag/pull/424

Is this another issue we should push upstream?

dg0yt replied:
So the whitespace changes are from that PR, and the patch will go away with the next release. Okay.
(Patches which exist only in vcpkg should be minimal. But temporary patches from upstream should be fine.)

JonLiu1993 commented:
@gotnone, Is work still being done for this PR?

JonLiu1993 commented:
Closing this PR since it seems that no progress is being made. Please ping us to reopen if work is still being done.
